libs/libskynet/src/seed.ts

```diff
@@ -0,0 +1,202 @@
+function deriveMyskyRootKeypair(userSeed: Uint8Array): Ed25519Keypair {
+  const saltBytes = new TextEncoder().encode("root discoverable key");
+  const saltHash = sha512(saltBytes);
+  const userSeedHash = sha512(userSeed);
+  const mergedHash = sha512(new Uint8Array([...saltHash, ...userSeedHash]));
+  const keyEntropy = mergedHash.slice(0, 32);
+
+  // Error is ignored because it should not be possible with the provided
+  // inputs.
+  const [keypair] = ed25519KeypairFromEntropy(keyEntropy);
+  return keypair;
+}
+
+// generateSeedPhraseDeterministic will generate and verify a seed phrase for
+// the user.
+function generateSeedPhraseDeterministic(password: string): [string, string | null] {
+  const u8 = new TextEncoder().encode(password);
+  const buf = sha512(u8);
+  const randNums = Uint16Array.from(buf);
+
+  // Generate the seed phrase from the randNums.
+  const seedWords = [];
+  for (let i = 0; i < SEED_ENTROPY_WORDS; i++) {
+    let wordIndex = randNums[i] % dictionary.length;
+    if (i == SEED_ENTROPY_WORDS - 1) {
+      wordIndex = randNums[i] % (dictionary.length / 4);
+    }
+    seedWords.push(dictionary[wordIndex]);
+  }
+
+  // Convert the seedWords to a seed.
+  const [seed, err1] = seedWordsToSeed(seedWords);
+  if (err1 !== null) {
+    return ["", err1];
+  }
+
+  // Compute the checksum.
+  const [checksumOne, checksumTwo, err2] = seedToChecksumWords(seed);
+  if (err2 !== null) {
+    return ["", err2];
+  }
+
+  // Assemble the final seed phrase and set the text field.
+  const allWords = [...seedWords, checksumOne, checksumTwo];
+  const seedPhrase = allWords.join(" ");
+  return [seedPhrase, null];
+}
+
+// seedToChecksumWords will compute the two checksum words for the provided
+// seed. The two return values are the two checksum words.
+function seedToChecksumWords(seed: Uint8Array): [string, string, string | null] {
+  // Input validation.
+  if (seed.length !== SEED_BYTES) {
+    return ["", "", `seed has the wrong length: ${seed.length}`];
+  }
+  // This line is just to pacify the linter about SEED_CHECKSUM_WORDS.
+  if (SEED_CHECKSUM_WORDS !== 2) {
+    return ["", "", "SEED_CHECKSUM_WORDS is not set to 2"];
+  }
+
+  // Get the hash.
+  const h = sha512(seed);
+
+  // Turn the hash into two words.
+  let word1 = h[0] << 8;
+  word1 += h[1];
+  word1 >>= 6;
+  let word2 = h[1] << 10;
+  word2 &= 0xffff;
+  word2 += h[2] << 2;
+  word2 >>= 6;
+  return [dictionary[word1], dictionary[word2], null];
+}
+
+// validSeedPhrase checks whether the provided seed phrase is valid, returning
+// an error if not. If the seed phrase is valid, the full seed will be returned
+// as a Uint8Array.
+function validSeedPhrase(seedPhrase: string): [Uint8Array, string | null] {
+  // Create a helper function to make the below code more readable.
+  const prefix = function (s: string): string {
+    return s.slice(0, DICTIONARY_UNIQUE_PREFIX);
+  };
+
+  // Pull the seed into its respective parts.
```

Comment:
I see you're using seed in different context
- seedPhrase - one string with all words
- seedWords - array of seed words
- seed - uint8array of seed words (no checksum)

now if that is correct then the comment here should say "pull the seed phrase into its respective parts"

it might sound like nitpick but this is already complex code and we should make sure comments and naming of variables is rock solid because it's going to be read by people who might not get the context or generalisation that's in place

I would also change "seed" to "seedBytes" (or "seed bytes" when used in comments) just to be verbose about the type since that is quite important wrt seed